### PR TITLE
 phase 0 — cover the redeem downside branch [BS-4845]

### PR DESCRIPTION
**Pull request by iamsahu**
## Description

This pull request introduces automated checks to ensure that all deployable production contracts remain within the EIP-170 runtime bytecode size limit, both in local development and continuous integration (CI). It adds a new script and CI step for Foundry builds, strengthens Hardhat configuration for stricter enforcement, and documents the rationale and usage of these safeguards. Additionally, it introduces new tests and documentation for contract behavior and formal verification coverage.

**Contract Size Enforcement:**

- Added a new script, `scripts/check_contract_sizes.sh`, which checks all deployable production contracts against the EIP-170 runtime bytecode limit during Foundry builds. This script is scoped to production contracts and allows for a configurable minimum margin.
- Integrated the new size check into the CI pipeline by adding a "Check contract sizes" step to `.github/workflows/Tests.yaml`, ensuring that oversized contracts are caught before deployment.
- Updated the `Makefile` to include a `size` target that runs the new size-check script, and marked it as a phony target. [[1]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L1-R1) [[2]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R22-R27)

**Hardhat Configuration Improvements:**

- Modified `hardhat.config.ts` to set `contractSizer.strict` to `true` for local and CI builds (except during coverage runs), causing Hardhat compilation to fail if any contract exceeds the EIP-170 limit. This ensures that the Hardhat build, which produces larger bytecode than Foundry, is also gated. [[1]](diffhunk://#diff-57cf5378f8cab20b02b279097ba6fad08eb53f747d81770045dda9c5f95effe3R19-R23) [[2]](diffhunk://#diff-57cf5378f8cab20b02b279097ba6fad08eb53f747d81770045dda9c5f95effe3R41-R47)

**Testing and Documentation:**

- Added two new tests to `RedeemManager.1.t.sol` to verify correct handling of redemption requests when the settlement rate falls below the request rate, and to ensure the cap logic behaves as intended in edge cases.
- Added a detailed `README.md` to `certora/` explaining the current state of formal verification, documenting known breakages, and outlining steps for future repairs.

## Notice

- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/liquid-collective/liquid-collective-protocol/pulls) for the same update/change?
- [ ] Have you assigned this PR to yourself?
- [ ] Have you added at least 1 reviewer?
- [ ] Have you updated the official documentation?
- [ ] Have you added sufficient documentation in your code?
- [ ] Have you added relevant tests to the official test suite?

## Pull Request Type

- [ ] 💫 New Feature (Breaking Change)
- [ ] 💫 New Feature (Non-breaking Change)
- [ ] 🛠️ Bug fix (Non-breaking Change: Fixes an issue)
- [ ] 🕹️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)

## Testing

- [ ] Have you tested this code with the official test suite?
- [ ] Have you tested this code manually?

### CODE REVIEW

File: certora/README.md
Comment (juliaaschmidt) [inline certora/README.md:1]: Is this meant as a friendly note to certora? Certora will fix this folder when they do the formal verification. I am not reviewing this file.

File: scripts/check_contract_sizes.sh [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+# Fails if any deployable PRODUCTION contract exceeds the EIP-170 runtime bytecode limit.
+#
+# Why this exists rather than plain `forge build --sizes`: that command exits non-zero on the
+# whole build, and two test-only harnesses (AccountingRiverV1, RiverV1ForceCommittable) are
+# already over the limit by design — they bolt extra helpers onto RiverV1. So the unscoped
+# command can never be used as a gate. This script scopes the check to contracts declared under
+# contracts/src (excluding interfaces/ and mock/), which is exactly the set that gets deployed.
+#
+# RiverV1 runs close to the limit and every cheap lever is already spent (see foundry.toml:
```
Comment (juliaaschmidt) [inline scripts/check_contract_sizes.sh]: I think these lines 11-20 can be removed. just AI waffle that will be outdated as soon as we change any of the configs.

File: scripts/check_contract_sizes.sh [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# which produces LARGER bytecode. A green run here is necessary but not sufficient — the hardhat
+# side is gated separately by `contractSizer.strict` in hardhat.config.ts.
+#
+# Usage:
+#   ./scripts/check_contract_sizes.sh
+#   MIN_RUNTIME_MARGIN=2048 ./scripts/check_contract_sizes.sh   # also fail if margin is too thin
+
+set -euo pipefail
+
+EIP170_LIMIT=24576
```
Comment (juliaaschmidt) [inline scripts/check_contract_sizes.sh]: can we add https://eips.ethereum.org/EIPS/eip-7954 size limit check too? It's now meant to be in Q4 2026

File: scripts/check_contract_sizes.sh [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+cd "$(dirname "$0")/.."
+
+# Names of every non-abstract contract and every library declared in production sources.
+# Internal-only libraries compile to ~16 byte stubs and pass trivially; LibOracleReporting is
+# the one that is genuinely deployed and worth watching.
+PRODUCTION_NAMES=$(
+  find contracts/src -name '*.sol' -not -path '*/interfaces/*' -not -path '*/mock/*' -print0 |
+    xargs -0 grep -hE '^(contract|library) [A-Za-z0-9_]+' |
+    sed -E 's/^(contract|library) ([A-Za-z0-9_]+).*/\2/' |
+    sort -u
```
Comment (zajck) [inline scripts/check_contract_sizes.sh]: Is getting only unique desired behaviour? At least technically, there could be two contracts with the same name, one over the limit and one not.  
I suppose we want to avoid duplicate names anyway, so in practice this should not be an issue, but then also dedup would not be necessary.

File: scripts/check_contract_sizes.sh [hunk 4]
```diff
@@ -0,0 +1,110 @@
+#! /bin/bash
+
+# Fails if any deployable PRODUCTION contract exceeds the EIP-170 runtime bytecode limit.
```
Comment (zajck) [inline scripts/check_contract_sizes.sh]: If we are checking EIP-170, we could also check [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860).
Not that we are close anywhere to hitting that, but `forge build --sizes` outputs both runtime and init size already.

File: certora/README.md
Comment (iamsahu) [reply certora/README.md:1]: Yup, we don't need a review on it.

File: scripts/check_contract_sizes.sh (hunk 1 above)
Comment (iamsahu) [reply scripts/check_contract_sizes.sh]: Agreed, removed

File: scripts/check_contract_sizes.sh (hunk 3 above)
Comment (iamsahu) [reply scripts/check_contract_sizes.sh]: I agree, dedup is not necessary, will remove that.

File: scripts/check_contract_sizes.sh (hunk 4 above)
Comment (iamsahu) [reply scripts/check_contract_sizes.sh]: Good idea, will add that check.

**Review by zajck-galaxy — APPROVED** [review]
LGTM

**Review by zajck — DISMISSED** [review]
LGTM

**Review by juliaaschmidt — DISMISSED** [review]

**Review by juliaaschmidt — DISMISSED** [review]

**Review by zajck — APPROVED** [review]

**Review by juliaaschmidt — APPROVED** [review]
